# 에러 처리

## Error객체

에러 객체가 갖는 프로퍼티

- message: 에러 메세지
- stack: 에러를 발생시킨 콜스택 호출 정보

### Error생성자 함수 종류

- Error
- SyntaxError
- ReferenceError - 참조할 수 없는 식별자
- RangeError
- URIError - encodeURI, decodeURI에 부적절한 인수 전달
- EvalError - eval함수의 에러 객체

호출된 에러를 catch하지 않으면 호출자 방향으로 전파된다.

그러나 setTime이나 프로미스 후속 처리 메서드의 콜백 함수는 호출자가 없다.

큐에 일시적으로 저장되었다가 콜스택이 비면 이벤트 루프에 의해 콜스택으로 푸시되어 실행되므로, 푸시된 콜백함수의 실행 컨텍스트는 콜스택의 가장 하부에 존재하며, 전파할 호출자가 존재하지 않는다.

## 프론트엔드에서의 에러 처리

React 16의 `Error Boundary`

카카오페이지 웹에서는 Fetcher컴포넌트로 API를 호출하여 데이터를 가져오고, Error Boundary로 비동기 컴포넌트에서의 에러를 처리함

Error Boundary없이 이런 형태로 에러에 대한 처리는 가능하지만, Error Boundary의 장점이 있다.

```jsx

  const { isLoading, error, data } = useSelector(state => state.productList);

  if (error) {
    return <div>문제가 발생했습니다.</div>;
  }
```

그러나 이 경우 api의 에러 뿐 아니라 이후 가공하며 에러가 발생하면 트리 전체가 마운트 해제되어 빈 화면을 볼 수 있다.

그러나 Error Boundary는 컴포넌트 렌더링 중에 발생한 에러를 모두 캐치하도록 동작한다. 해당 boundary에서 에러 처리가 다시 실패될 경우 다음 에러 바운더리에서 잡히도록 동작한다

카카오에서는 Fetcher에서는 에러를 throw만 하고, catch는 Error boundary에서 동작하여 에러 영역을 제한한다.
